pkg/prometheus/agent/operator.go

```diff
@@ -0,0 +1,1404 @@
+			c.rr.EnqueueForReconciliation(p)
+			return
+		}
+
+		// Check for Prometheus instances selecting PodMonitors in the NS.
+		pmNSSelector, err := metav1.LabelSelectorAsSelector(p.Spec.PodMonitorNamespaceSelector)
+		if err != nil {
+			level.Error(c.logger).Log(
+				"msg", fmt.Sprintf("failed to convert PodMonitorNamespaceSelector of %q to selector", p.Name),
+				"err", err,
+			)
+			return
+		}
+
+		if pmNSSelector.Matches(labels.Set(ns.Labels)) {
+			c.rr.EnqueueForReconciliation(p)
+			return
+		}
+
+		// Check for Prometheus instances selecting Probes in the NS.
+		bmNSSelector, err := metav1.LabelSelectorAsSelector(p.Spec.ProbeNamespaceSelector)
+		if err != nil {
+			level.Error(c.logger).Log(
+				"msg", fmt.Sprintf("failed to convert ProbeNamespaceSelector of %q to selector", p.Name),
+				"err", err,
+			)
+			return
+		}
+
+		if bmNSSelector.Matches(labels.Set(ns.Labels)) {
+			c.rr.EnqueueForReconciliation(p)
+			return
+		}
+	})
+	if err != nil {
+		level.Error(c.logger).Log(
+			"msg", "listing all Prometheus instances from cache failed",
+			"err", err,
+		)
+	}
+
+}
+
+func (c *Operator) handleMonitorNamespaceUpdate(oldo, curo interface{}) {
+	old := oldo.(*v1.Namespace)
+	cur := curo.(*v1.Namespace)
+
+	level.Debug(c.logger).Log("msg", "update handler", "namespace", cur.GetName(), "old", old.ResourceVersion, "cur", cur.ResourceVersion)
+
+	// Periodic resync may resend the Namespace without changes
+	// in-between.
+	if old.ResourceVersion == cur.ResourceVersion {
+		return
+	}
+
+	level.Debug(c.logger).Log("msg", "Monitor namespace updated", "namespace", cur.GetName())
+	c.metrics.TriggerByCounter("Namespace", operator.UpdateEvent).Inc()
+
+	// Check for Prometheus Agent instances selecting ServiceMonitors, PodMonitors,
+	// and Probes in the namespace.
+	err := c.promInfs.ListAll(labels.Everything(), func(obj interface{}) {
+		p := obj.(*monitoringv1alpha1.PrometheusAgent)
+
+		for name, selector := range map[string]*metav1.LabelSelector{
+			"PodMonitors":     p.Spec.PodMonitorNamespaceSelector,
+			"Probes":          p.Spec.ProbeNamespaceSelector,
+			"ServiceMonitors": p.Spec.ServiceMonitorNamespaceSelector,
+		} {
+
+			sync, err := k8sutil.LabelSelectionHasChanged(old.Labels, cur.Labels, selector)
+			if err != nil {
+				level.Error(c.logger).Log(
+					"err", err,
+					"name", p.Name,
+					"namespace", p.Namespace,
+					"subresource", name,
+				)
+				return
+			}
+
+			if sync {
+				c.rr.EnqueueForReconciliation(p)
+				return
+			}
+		}
+	})
+	if err != nil {
+		level.Error(c.logger).Log(
+			"msg", "listing all Prometheus Agent instances from cache failed",
+			"err", err,
+		)
+	}
+}
+
+// checkAgentAuthorization checks if the operator has access to PrometheusAgent CRD
+// It checks individual namespaces if an allowlist was provided, otherwise it checks cluster-wide.
+func (c *Operator) checkAgentAuthorization(ctx context.Context) (bool, error) {
+	if c.config.Namespaces.AllowList != nil {
```

Comment:
it needs to check for `c.config.Namespaces.AllowListc.config.Namespaces.PrometheusAllowList`.
Also the map is never nil or empty. When watching on all namespaces, the map contains a single key equal to `v1.NamespaceAll`.
https://github.com/prometheus-operator/prometheus-operator/blob/d16abbd7c1d206f848911e22c4190211fe4d2ae8/cmd/operator/main.go#L230

Reply:
Oh interesting, I didn't know that :)

`struct{}{}`, when cast into a string, becomes an empty string `""`. Which is exactly what SelfSubjectAccessReview expects to perform a cluster-wide review. Just perfect!

Reply:
Let me ask, what is the difference between `AllowList` and `PrometheusAllowList`?

Reply:
AllowList is the list of namespaces where the operator should watch for all resources except Prometheus, Alertmanager, AlertmanagerConfig and ThanosRuler. Which means ServiceMonitor, PodMonitor, Probe and PrometheusRule objects.
https://github.com/prometheus-operator/prometheus-operator/blob/d16abbd7c1d206f848911e22c4190211fe4d2ae8/pkg/operator/config.go#L109-L114

The operator main() function will default the *AllowList fields to AllowList if they are empty:

https://github.com/prometheus-operator/prometheus-operator/blob/d16abbd7c1d206f848911e22c4190211fe4d2ae8/cmd/operator/main.go#L239-L253